### PR TITLE
Avoid iterator invalidation without effective insert

### PR DESCRIPTION
**Pull request by artem**
There's an issue with accessing an existing value with operator[] when
hashtable size reaches a certain value: iterator is sometimes being
invalidated leading to broken iterators invalidation contract:

> - insert, emplace, emplace_hint, operator[]: if there is an effective
> insert, invalidate the iterators.

or use-after-free from calling code if user is unaware of such behaviour.

Fix this issue by actually issuing a rehash in effective erase and insert
routines.

### CODE REVIEW

**Comment by artem** [comment]
PS: docs also say that `erase()` always invalidates the iterators, though the current implementation does that only on effective erase. Should the doc be adjusted as well?

File: include/tsl/sparse_hash.h [hunk 1]
```diff
@@ -1929,6 +1936,10 @@ class sparse_hash : private Allocator,
           m_nb_elements--;
           m_nb_deleted_buckets++;
 
+          if (size() + m_nb_deleted_buckets >= m_load_threshold_clear_deleted) {
+            clear_deleted_buckets();
```
Comment (Tessil) [inline include/tsl/sparse_hash.h]: `clear_deleted_buckets` can throw an exception in some cases. A element could thus be correctly erased but the `erase` method could thus throw an exception which goes against the exception safety of the map.

**Review by Tessil — CHANGES_REQUESTED** [review]

File: include/tsl/sparse_hash.h [hunk 2]
```diff
@@ -1833,13 +1833,9 @@ class sparse_hash : private Allocator,
   template <class K, class... Args>
   std::pair<iterator, bool> insert_impl(const K &key,
                                         Args &&...value_type_args) {
-    if (size() >= m_load_threshold_rehash) {
+    if (m_sparse_buckets_data.empty()) {
```
Comment (Tessil) [inline include/tsl/sparse_hash.h]: These lines may not be needed. An empty `m_sparse_buckets_data` means that `m_sparse_buckets` points to `static_empty_sparse_bucket_ptr` and we will thus reach the later rehash call (just a bit slower on the first insertion but worth it for the simplicity).

**Review by Tessil — CHANGES_REQUESTED** [review]

**Comment by Tessil** [comment]
Thank you, we effectively should avoid rehashing when no actual insert is done. 

I added some comments to the PR code.

**Comment by artem** [comment]
Thanks for the feedback. I have pushed fixes for review and additionally reverted `+ 1` in load threshold check to preserve original behaviour

File: tests/sparse_map_tests.cpp [hunk 3]
```diff
@@ -1132,6 +1132,11 @@ BOOST_AUTO_TEST_CASE(
   BOOST_CHECK_EQUAL(map.bucket_count(), 64);
   BOOST_CHECK_EQUAL(map.max_load_factor(), 0.8f);
 
+  /**
+   * Current map load threshold for rehashing is expected to be 52 elements.
```
Comment (Tessil) [inline tests/sparse_map_tests.cpp]: The load rehash threshold is 51 not 52. I think though that the comment can eventually be removed.

**Comment by Tessil** [comment]
Thank you for the change, looks good. Just a small comment left.

Note that the map doesn't guarantee that erase is `noexcept` per the commit message. The problem is more that an exception could be thrown during the rehash after an element was successfully erased which goes against the strong exception guarantee (see https://en.cppreference.com/w/cpp/language/exceptions). 

I also would like to limit the rehash to insertions as done in `std::unordered_map`

**Comment by artem** [comment]
I think I have addressed all of your comments in the last revision and it is safe to merge. Also, I have squashed all the commits into a single one (since they no longer make sense on their own).

Sorry for the rather long delay, been busy with other things.

**Review by Tessil — APPROVED** [review]

**Comment by Tessil** [comment]
Thank you, it looks good, I merged the change.

**Pull request merged**